### ISSUE TITLE
3rd Party API request. Having trouble figuring out how to parse response data and store locally.

### ISSUE THREAD

**Issue by louarnos** [ga-wdi-boston/capstone-project#154]
I currently have my backend set so that a post request to route '/eventful' will fire a request to a 3rd party API. I have managed to get be able to get this data to save as a single entry in the data store on the front end. My issue is that this data comes back as an array of events. 

Because this data is not present in my database prior to a user sending requesting it from the api, a model hook does not have anything to work with on page load and throws an error (or so it seems to me);

I've been trying to figure out how to take the single entry (the array of events) and to create new records with it so that I can then render them. Here is my code to store the single entry:

My eventful Route:

``` bash
import Ember from 'ember';

export default Ember.Route.extend({
  // model () {
  //   return this.get('store').find('eventful');
  // },

  data : {},

  ajax: Ember.inject.service(),

  actions: {
    searchForEvents (data) {
      let eventful = this.get('store').createRecord('eventful', data);
      return eventful.save();
      // return this.get('store').findAll('eventful');
      // let something = this.get('store').findAll('eventful');
      // console.log('something:', something);
      // return eventful.save();
    }
  }
});
```

My eventful adapter:

``` bash
import ApplicationAdapter from 'ga-wdi-boston.event-bookmarker/application/adapter';
import Ember from 'ember';

export default ApplicationAdapter.extend({
  ajax: Ember.inject.service(),
  createRecord (store, type, record) {
  let serialized = this.serialize(record, { includeId: true});
  let data = { serialized };
  return this.get('ajax').request('/eventful', {
    method: 'POST',
    data: data,
  });
  // .then((events) => {
  //   // console.log(events);
  //   events.eventful[0].event.forEach(function (concert){
  //     // console.log(concert);
  //     return concert;
  //     // let eventfulConcert = this.get('store').createRecord('eventful-event', concert);
  //     // return eventfulConcert.save();
  //     });
  //   });
  }
});
```

My eventful Model:

``` bash
import DS from 'ember-data';

export default DS.Model.extend({
  event: DS.attr('string'),
});
```

As you can see in the lines commented out in my adapter, I have tried to do a forEach on the array that is returned from the request to the 3rd party api and then save each individual event as a different model. I get an error when i do this that says: 

```
rest.js:390 Uncaught TypeError: this.ajax is not a functionfindAll @ rest.js:390_findAll @ finders.js:136_fetchAll @ store.js:909findAll @ store.js:881searchForEvents @ route.js:17triggerEvent @ ember.debug.js:28580trigger @ ember.debug.js:53473Router.trigger @ ember.debug.js:52232_emberRuntimeSystemObject.default.extend.send @ ember.debug.js:28045_emberMetalMixin.Mixin.create.send @ ember.debug.js:32723_emberMetalMixin.Mixin.create.triggerAction @ ember.debug.js:36297_emberViewsViewsView.default.extend.sendAction @ ember.debug.js:41731searchForEvents @ component.js:6_emberViewsViewsView.default.extend.send @ ember.debug.js:41747superWrapper @ ember.debug.js:23110runRegisteredAction @ ember.debug.js:29633Backburner.run @ ember.debug.js:678run @ ember.debug.js:21139actions.push.handler @ ember.debug.js:29627(anonymous function) @ ember.debug.js:44705jQuery.event.dispatch @ jquery.js:4737elemData.handle @ jquery.js:4549
```

I have done some research, and it seems I may have to define different DS.attr() to handle arrays. My backend has two different embed schemas in users (events and followees) so I'm going to have to learn to handle storing data as arrays in the ember store.

**Comment by RealWeeks** [comment]
So front end request to /eventful triggers an api request on your express backend that performs an request that returns an array. The payload is then sent from the API to the Ember frontend?

**Comment by RealWeeks** [comment]
have you tried to parse the array on the backend or store the payload to your db?

**Comment by gaand** [comment]
Do you know what's causing the `this.ajax is not a function` error?  That seems like it might be the first thing you need to fix.

**Comment by louarnos** [comment]
I can successfully store the data on the ember front end. I'm working on adjusting how the data is coming back from the api. @gaand I'm not sure what is causing this.ajax to not be a function. I'll look into it once I sort a few other things out.

**Comment by louarnos** [comment]
I've realized that the code in my adapter is not sending the data I want. I am getting data back, but not what I expect. I'm not sure how to pass the data I want. I've tried moving my ajax call into my model instead of my adapter, but when I try to save event record then, it pluralizes it and give me a 404.

Not exactly sure how to approach my confusion. Considering dropping ember all together.

**Comment by RealWeeks** [comment]
Is the data your getting back not formatted how you expect or not at all the data you expected.  IE: do you need to manipulate it or is it something else?

**Comment by louarnos** [comment]
I figured out how to send what I want. So now i am getting what I expect. It isnt correctly formatted but Ill deal with that later. I need to find a way to access the data that is stored and then I'll worry about formatting... Currently whenever I try to 

```
this.get('store').findAll('eventful')
```

in my eventful model I get an error saying this. ajax is not a function

**Comment by RealWeeks** [comment]
what have you done so far for correct the this.ajax error?

**Comment by louarnos** [comment]
@J-Weeks I'm not getting this error anymore. I am currently using a custom DS.attr('array') but i cant figure out how to make use of the data.

**Comment by RealWeeks** [comment]
Can you be a bit more specific with the issue you're having?

**Comment by louarnos** [comment]
@J-Weeks Its difficult to explain. I have data in the store, but I cant access it. I think it maybe how its structured.

**Comment by louarnos** [comment]
@J-Weeks I think I need to use a jsonSerializer or make a transform for nested objects

**Comment by RealWeeks** [comment]
@louarnos so where are you at with this?

**Comment by louarnos** [comment]
@J-Weeks The data returned from the API is much larger than I need. I am working on making a model for it and then skimming it down on the backend. 

I am afraid I am not on track and am wasting time on the wrong things. I'd appreciate being able to talk to someone to so that I am reassured that I am approaching the problem in the right way.

**Comment by RealWeeks** [comment]
@louarnos I feel like your problem has shifted throughout this thread. Before myself or @gaand assigns can you summarize the problem that you're currently having. I'm not trying to slow you down or be a jerk but I think it would help you if you could summarize the current issue and the specific bug you're trying to tackle.

**Comment by louarnos** [comment]
@J-Weeks It has been shifting. I was down a road that didn't make sense and I have had to backtrack. 

So at this point I have the 3rd party api returning an array  objects to the front end. My issue is I dont know how to handle and array of objects in ember. There is not DS.attr that I see working well.

I know I need to strip the data down on the backend to only the keys I'm using, but once I have that done I dont know how I will handle this format of data. I have some ideas about how i want to approach it, but I've spent about 24 hours on getting this data displayed, so I'm hesitant to invest a lot of time in something I'm not convinced is the right approach.

**Comment by RealWeeks** [comment]
have you looked into the Ember array module http://emberjs.com/api/classes/Ember.Array.html

**Comment by gaand** [comment]
Why grab it in the client?  Why not grab it in your api end and return what you need to the front end?

Store what you need and just send the relevant results?

**Comment by louarnos** [comment]
I guess I'm confused at what you mean exactly @gaand. My the request returns an array of concert information, and i need to get it to the frontend. So the payload is an array, even if I strip down to just the keys I need. 

@J-Weeks Is that a DS.attr() type? I'm looking for a way to model the data coming from the API.

**Comment by RealWeeks** [comment]
it's an ember.enum I was looking though the docs for ember arrays and using them with ember objects.  Read this re: your earlier problem with constructing Ember arrays https://guides.emberjs.com/v1.10.0/enumerables/

**Comment by gaand** [comment]
I'm talking about the streaming solution I suggested for your group project for an API with a key (it might not apply).

**Comment by louarnos** [comment]
@gaand hmm haven't considered it.

**Comment by louarnos** [comment]
@J-Weeks ok, my back end is returning data from the 3rd party in the form I would like. I am having trouble creating records for each event in the array. Currently I have an 'eventful' route that makes the request to the API and i have the eventful-event model that models each event. After the resolution of the api request promise I execute a foreach on the payload and try to create instances on eventful-events in my DS. Here is my code:

```
import ApplicationAdapter from 'ga-wdi-boston.event-bookmarker/application/adapter';
import Ember from 'ember';

export default ApplicationAdapter.extend({
  ajax: Ember.inject.service(),

  createRecord (store, type, record) {
    let serialized = this.serialize(record, { includeId: true});
    let data = { serialized };
    let dataToSend = {'keywords': record.record.keywords, 'location': record.record.location };
    return this.get('ajax').request('/eventful', {
      method: 'POST',
      data: dataToSend,
    }).then((events) => {
      events['eventful-event'].forEach(function(concert){
        this.get('store').createRecord('eventful-event', concert);
      });
    });
  }
});
```

I get the error (which seems to be plaguing me): 

```
ember.debug.js:32096 TypeError: Cannot read property 'get' of undefined
```

**Comment by RealWeeks** [comment]
what is the value of "this"

**Comment by louarnos** [comment]
@J-Weeks undefined...

**Comment by louarnos** [comment]
I'm not sure if I should even be doing this in my adapter. I've been trying to render this data for a long time.

**Comment by louarnos** [comment]
yayyyy

**Comment by louarnos** [comment]
it works

**Comment by louarnos** [comment]
```
store.createRecord('eventful-event', concert)

```

fixed the issue.
